packages/storage/amplify_storage_s3_dart/lib/src/amplify_storage_s3_dart_io.dart

```diff
@@ -0,0 +1,199 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:amplify_storage_s3_dart/amplify_storage_s3_dart.dart';
+import 'package:amplify_storage_s3_dart/src/amplify_storage_s3_dart_base.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+
+/// Symbol for testing.
+@visibleForTesting
+const testEnsureDestinationWritableOverride =
+    #_testEnsureDestinationWritableOverride;
+
+/// Symbol for testing.
+@visibleForTesting
+const testTempFileOverride = #_testTempFileOverride;
+
+/// Symbol for testing.
+@visibleForTesting
+const testDestinationOverride = #_testDestinationOverride;
+
+Future<String> Function(AWSFile)? get _ensureDestinationWritableOverride =>
+    Zone.current[testEnsureDestinationWritableOverride] as Future<String>
+        Function(AWSFile)?;
+
+File? get _tempFileOverride => Zone.current[testTempFileOverride] as File?;
+
+File? get _destinationOverride =>
+    Zone.current[testDestinationOverride] as File?;
```

Comment:
This is probably overuse of this - I think I have a total of 3-4 of these throughout all my packages. As you can see, it becomes very hard to keep track of these, especially when they're platform-specific. We should be preferring dependency injection over this.

If you want to mock files, then I would mock+inject the full filesystem with `package:file`: https://pub.dev/packages/file. If it's too much work to implement, then just do integration tests. It's not worth the benefits of some unit tests.